File: src/lib/pilot/rag/confidence.ts
import type { ClassificationResult, RetrievalHit } from "../types";

export function calculateConfidence(args: {
  hits: RetrievalHit[];
  citedSentenceRatio: number;
  classification: ClassificationResult;
  insufficientEvidence: boolean;
  securitySensitive: boolean;
}): number {
  if (args.insufficientEvidence) {
    return Math.min(0.34, 0.18 + (args.hits[0]?.score ?? 0) * 0.2);
  }
  const top = args.hits[0]?.score ?? 0;
  const mean = args.hits.length
    ? args.hits.reduce((s, h) => s + h.score, 0) / args.hits.length
    : 0;
  let conf = 0.25 + top * 0.45 + mean * 0.15 + args.citedSentenceRatio * 0.2;
  if (args.securitySensitive) conf = Math.min(conf, 0.58);
  if (args.classification.safetyFlags.includes("prompt_injection_in_customer_message")) {
    conf = Math.min(conf, 0.4);
  }
  return Math.max(0.05, Math.min(0.97, round2(conf)));
}

export function round2(n: number): number {
  return Math.round(n * 100) / 100;
}

File: backend/app/rag/confidence.py
"""Draft confidence scoring."""

from __future__ import annotations

from typing import Any


def round2(n: float) -> float:
    return round(n * 100) / 100


def calculate_confidence(
    *,
    hits: list[dict[str, Any]],
    cited_sentence_ratio: float,
    classification: dict[str, Any],
    insufficient_evidence: bool,
    security_sensitive: bool,
) -> float:
    if insufficient_evidence:
        top = hits[0]["score"] if hits else 0.0
        return min(0.34, 0.18 + float(top) * 0.2)
    top = hits[0]["score"] if hits else 0.0
    mean = sum(float(h["score"]) for h in hits) / len(hits) if hits else 0.0
    conf = 0.25 + float(top) * 0.45 + mean * 0.15 + cited_sentence_ratio * 0.2
    if security_sensitive:
        conf = min(conf, 0.58)
    flags = classification.get("safety_flags") or classification.get("safetyFlags") or []
    if "prompt_injection_in_customer_message" in flags:
        conf = min(conf, 0.4)
    return max(0.05, min(0.97, round2(conf)))
